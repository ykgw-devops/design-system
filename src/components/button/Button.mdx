---
name: Button
menu: Components
---

import { Playground, Props } from 'docz'
import Button from './Button'
import Group from './Group'
import Icon from '../icon/Icon'
import { Popup } from '../popup/Popup'
import Segment from '../segment/Segment'

# Button

## Basic usage

<Playground>
  <Button>Primary</Button>

  <Button>
    <Icon name='add_shopping_cart' /> Add to cart
  </Button>

  <Button outline kind='danger'>
    <Icon name='stop' /> Stop recording
  </Button>
</Playground>

## As link

<Playground>
  <Button as='a' href='https://google.com/' target='_blank'>Google <Icon name='launch' /></Button>
</Playground>

## Kind

<Playground>
  <Button kind="primary">Primary</Button>
  <Button kind="secondary">Secondary</Button>
  <Button kind="warning">Warning</Button>
  <Button kind="danger">Danger</Button>
  <Button kind="success">Success</Button>
</Playground>

## Pill

<Playground>
  <Button pill kind="danger">Red Pill</Button>
  <Button pill kind="primary">Blue Pill</Button>
</Playground>

## Outline

<Playground>
  <Button kind="primary" outline>Primary</Button>
  <Button kind="secondary" outline>Secondary</Button>
  <Button kind="warning" outline>Warning</Button>
  <Button kind="danger" outline>Danger</Button>
  <Button kind="success" outline>Success</Button>
</Playground>

## Disabled

<Playground>
  <Button disabled>Disabled</Button>

  <Button pill outline kind="warning" disabled>Disabled Warning Button</Button>
</Playground>

## Size

<Playground>
  <Button size='tiny'>Tiny</Button>

  <Button size='small'>Small</Button>

  <Button size='medium'>Medium</Button>

  <Button size='large'>Large</Button>
</Playground>

## Loading

<Playground>
  <Button loading>Hold on tight...</Button>
  <Button kind='success' loading>Hold on tight...</Button>
  <Button kind='danger' loading>Hold on tight...</Button>
  <br />
  <Button disabled loading>Hold on tight...</Button>
  <Button disabled kind='success' loading>Hold on tight...</Button>
  <Button disabled kind='danger' loading>Hold on tight...</Button>
  <br />
  <Button outline loading>Hold on tight...</Button>
  <Button outline kind='success' loading>Hold on tight...</Button>
  <Button outline kind='danger' loading>Hold on tight...</Button>
  <br />
  <Button disabled outline loading>Hold on tight...</Button>
  <Button disabled outline kind='success' loading>Hold on tight...</Button>
  <Button disabled outline kind='danger' loading>Hold on tight...</Button>
</Playground>

## Properties

<Props of={Button}></Props>

# Group

<Playground>
  Base:
  <Button.Group>
    <Button>
      <Icon name='filter_1' /> Option #1
    </Button>
    <Button outline>Option #2</Button>
    <Button>
      <Icon name='filter_3' /> Option #3
    </Button>
  </Button.Group>

  <br />

  Dropdown Button:
  <Button.Group>
    <Button>
      <Icon name='filter_1' /> Option #1
    </Button>
    <Popup
      placement='bottom-end'
      appendTo={() => document.body}
      content={
        <Segment>
          Some Dropdown Content Here
        </Segment>
      }
    >
      <Button>
        <Icon name='keyboard_arrow_down' />
      </Button>
    </Popup>
  </Button.Group>

  <br />

  <Button.Group>
    <Button kind='secondary'>
      <Icon name='filter_1' /> Option #1
    </Button>
    <Button kind='secondary'>Option #2</Button>
    <Button outline kind='secondary'>
      <Icon name='filter_3' /> Option #3
    </Button>
  </Button.Group>

  <br />

  Compact width:
  <Button.Group compact>
    <Button>Number #1</Button>
    <Button kind='danger'>Number #2</Button>
    <Button kind='success'>Number #3</Button>
  </Button.Group>

  <br />

  Fluid:
  <Button.Group fluid>
    <Button>Text</Button>
    <Button kind='danger'>Text</Button>
    <Button kind='success'>How much wood would a woodchuck chuck if a woodchuck could chuck wood?</Button>
  </Button.Group>

  <br />

  Equal width:
  <Button.Group equalWidth>
    <Button>Number #1</Button>
    <Button kind='danger'>Number #2</Button>
    <Button kind='success'>How much wood would a woodchuck chuck if a woodchuck could chuck wood?</Button>
  </Button.Group>
</Playground>

## Properties

<Props of={Group}></Props>
