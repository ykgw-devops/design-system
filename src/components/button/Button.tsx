import React, { ButtonHTMLAttributes } from 'react'
import styled from '@emotion/styled'
import isPropValid from '@emotion/is-prop-valid'
import { setDisplayName } from 'recompose'

import Group from './Group'
import Loader from '../loader/Loader'

import { base, kind, outline, pill, disabled, size, colorFromProps } from './Button.styles'

type ButtonKind = 'primary' | 'secondary' | 'warning' | 'danger' | 'success'
type ButtonSize = 'tiny' | 'small' | 'medium' | 'large'

export interface IButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  outline?: boolean;
  pill?: boolean;
  disabled?: boolean;
  loading?: boolean;
  kind?: ButtonKind;
  size?: ButtonSize;
}

const ValidButton = styled('button', { shouldForwardProp: isPropValid })

const Button = ValidButton<IButtonProps>`
  ${base};
  ${props => kind(props)};
  ${props => size(props)};
  ${props => props.outline && outline(props)};
  ${props => props.pill && pill};
  ${props => props.disabled && disabled(props)};
`

const EnhancedButton = React.forwardRef((props: IButtonProps, ref: React.Ref<any>) => {
  const { loading, children, ...rest } = props

  return (
    <Button {...rest} ref={ref}>
      {loading && (
        <>
          <Loader
            color={colorFromProps(props)}
            size={14}
            thickness={2}
          />
          &nbsp;
        </>
      )}
      {children}
    </Button>
  )
})

EnhancedButton.Group = setDisplayName('Button.Group')(Group)

EnhancedButton.defaultProps = {
  type: 'button',
  kind: 'primary',
  size: 'medium',
  pill: false,
  outline: false,
  disabled: false,
  loading: false
}

export default EnhancedButton
