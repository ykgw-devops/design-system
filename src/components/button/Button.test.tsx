import '@testing-library/jest-dom'
import React from 'react'
import { render, fireEvent, screen } from '@testing-library/react'
import serializer from 'jest-emotion'

import Button from './Button'

expect.addSnapshotSerializer(serializer)

test('button DOM', () => {
  const { asFragment } = render(
    <Button>Hello, World!</Button>
  )

  expect(asFragment()).toMatchSnapshot()
})

test('regular button', () => {
  render(
    <Button>Hello, World!</Button>
  )

  expect(screen.getByRole('button')).toHaveTextContent('Hello, World!')
  expect(screen.getByRole('button')).toHaveAttribute('type', 'button')
})

test('button with kind', () => {
  render(
    <Button kind='secondary'>Hello, World!</Button>
  )

  expect(screen.getByRole('button')).toBeInTheDocument()
})


test('disabled button', () => {
  render(
    <Button disabled>Hello, World!</Button>
  )
  expect(screen.getByRole('button')).toBeDisabled()
})

test('click button', () => {
  const onClick = jest.fn()

  render(
    <Button onClick={onClick}>Hello, World!</Button>
  )

  fireEvent.click(screen.getByRole('button'))
  expect(onClick).toHaveBeenCalledTimes(1)
})
