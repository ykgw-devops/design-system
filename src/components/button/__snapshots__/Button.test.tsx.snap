// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button DOM 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  white-space: pre;
  text-align: center;
  background-color: #1E90FF;
  border-radius: 0.333333rem;
  border: solid 1px #1b81e5;
  color: #fff;
  cursor: pointer;
  font-family: "Fira Sans",system-ui,sans-serif;
  font-weight: 400;
  padding: 0.65em 1em;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: background 80ms ease-in;
  transition: background 80ms ease-in;
  vertical-align: middle;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background-color: #1E90FF;
  color: white;
  border-color: #1b81e5;
  font-size: 0.9rem;
  line-height: 0.9rem;
}

.emotion-0:hover {
  background-color: #1b81e5;
}

.emotion-0:hover {
  background-color: #1b81e5;
}

<button
    class="emotion-0 emotion-1"
    kind="primary"
    type="button"
  >
    Hello, World!
  </button>
</DocumentFragment>
`;
